pkg/query-service/app/queryBuilder/query_builder.go

```diff
@@ -320,7 +320,46 @@ func (c *cacheKeyGenerator) GenerateKeys(params *v3.QueryRangeParamsV3) map[stri
 
 	// Build keys for each builder query
 	for queryName, query := range params.CompositeQuery.BuilderQueries {
-		if query.Expression == queryName && query.DataSource == v3.DataSourceMetrics {
+		if query.Expression == queryName && query.DataSource == v3.DataSourceLogs {
+			var parts []string
+
+			// We need to build uniqe cache query for BuilderQuery
+			parts = append(parts, fmt.Sprintf("source=%s", query.DataSource))
+			parts = append(parts, fmt.Sprintf("step=%d", query.StepInterval))
+			parts = append(parts, fmt.Sprintf("aggregate=%s", query.AggregateOperator))
+			parts = append(parts, fmt.Sprintf("limit=%d", query.Limit))
```

Comment:
Ideally, the limit and order by etc... shouldn't be part of the cache key because they don't change the main query but the end result processing. Please create an issue to move this post-processing logic to query-service. We already have things in place to do this today but you address it separately.

Reply:
we need it for logs

ex:- order by route asc with limit 1 and order by route desc with limit 1 will result in different response.

For logs we generate different queries for the both above cases.

Reply:
added to https://github.com/SigNoz/signoz/issues/4527